compiler-core/templates/prelude.mjs

```diff
@@ -138,21 +138,53 @@ export class UtfCodepoint {
 
 // @internal
 export function toBitArray(segments) {
-  let size = (segment) =>
-    segment instanceof Uint8Array ? segment.byteLength : 1;
-  let bytes = segments.reduce((acc, segment) => acc + size(segment), 0);
-  let view = new DataView(new ArrayBuffer(bytes));
+  if (segments.length === 0) {
+    return new BitArray(new Uint8Array());
+  }
+
+  if (segments.length === 1) {
+    // When there is a single Uint8Array segment, pass it directly to the bit
+    // array constructor to avoid a copy
+    if (segments[0] instanceof Uint8Array) {
+      return new BitArray(segments[0]);
+    }
+
+    return new BitArray(new Uint8Array(segments));
+  }
+
+  // Count the total number of bytes, and check if there are any Uint8Array
+  // segments
+  let bytes = 0;
+  let hasUint8ArraySegment = false;
+  for (const segment of segments) {
+    if (segment instanceof Uint8Array) {
+      bytes += segment.byteLength;
+      hasUint8ArraySegment = true;
+    } else {
+      bytes++;
+    }
+  }
+
+  // If there aren't any Uint8Array segments then pass the segments array
+  // directly to the Uint8Array constructor
+  if (!hasUint8ArraySegment) {
+    return new BitArray(new Uint8Array(segments));
+  }
+
+  // Copy the segments into a Uint8Array
+  let u8Array = new Uint8Array(bytes);
   let cursor = 0;
   for (let segment of segments) {
     if (segment instanceof Uint8Array) {
-      new Uint8Array(view.buffer).set(segment, cursor);
+      u8Array.set(segment, cursor);
       cursor += segment.byteLength;
     } else {
-      view.setInt8(cursor, segment);
+      u8Array[cursor] = segment;
```

Comment:
What is this doing?

Reply:
This is handling a segment that's a `number` when the input segments contain both numbers and `Uint8Array`s, e.g. when constructing something like `<<"A", 123>>`.

Previously `DataView.setInt8()` was being used, but now a numeric `segment` value is set directly on the final `Uint8Array`. If `segment` is outside [0 - 255] it will be wrapped into that range by the assignment, similar to how previously `segment` was being wrapped into the [-128 - 127] range by `DataView.setInt8()`.

The end result is the same, and also the updated tests now explicitly cover all of the integer wraparound points.